### PR TITLE
Exception handling for IHostedService implementations in start/stop methods

### LINKED ISSUE #4891: Revisit: IHostedServices can now throw exceptions on Start/Stop in 3.0
This behavior changed in .NET Core 3.0: https://github.com/aspnet/AspNetCore/pull/7657.

We had one test (`StandbyModeE2E_LinuxContainer`) that was failing at startup because the python language worker wasn't present. Our test in 2.0 passed happily because the host would swallow this exception. Now it failed. I've wrapped the startup in a try/catch, but we need to re-evaluate how we want to handle this if StartAsync() fails in our hosted services.

### PR DESCRIPTION
**Pull request by surgupta-msft**
### Issue describing the changes in this PR
resolves #4891 

The PR adds try-catch block in start/stop methods of all the implementations of IHostedService and IManagedHostedService.

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerActivityPublisher.cs [hunk 1]
```diff
@@ -75,27 +75,41 @@ private void OnStandbyOptionsChange()
...
+            try
+            {
+                _logger.LogInformation($"Stopping {nameof(LinuxContainerActivityPublisher)}");
 
-            // stop the timer if it has been started
-            _timer?.Change(Timeout.Infinite, Timeout.Infinite);
+                // stop the timer if it has been started
+                _timer?.Change(Timeout.Infinite, Timeout.Infinite);
+            }
+            catch (Exception ex)
```
Comment (kshyju) [inline src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerActivityPublisher.cs:109]: When the cancellation token is cancelled (_as a result of the infra code is signaling that the app is shutting down_), Would it throw `TaskCanceledException` or `OperationCanceledException` ?. Should we exclude that case before logging it as an error? 

Something like
````
catch (Exception ex)
{
    if (!cancellationToken.IsCancellationRequested)
    {
      //log error
    }
}
````

Comment (kshyju) [inline src/WebJobs.Script.WebHost/ContainerManagement/LinuxContainerActivityPublisher.cs:109]: When the cancellation token is cancelled (_as a result of the infra code is signaling that the app is shutting down_), it is going to throw `TaskCanceledException` or `OperationCanceledException`. Should we exclude that case before logging it as an error? 

Something like
````
catch (Exception ex)
{
    if (!cancellationToken.IsCancellationRequested)
    {
      //log error
    }
}
````

**Review by brettsam — COMMENTED** [review]
Looking at this -- I think that we only want to try/catch the startup of services that are optional. In a lot of cases, the host cannot really run without some of these running and it should fail to startup if an exception is thrown.

In fact, the only one that I think should have any try/catch around its startup is the `AssemblyAnalysisService`.  I think we should backout the changes from all the others.

**Comment by liliankasem** [comment]
> Looking at this -- I think that we only want to try/catch the startup of services that are optional. In a lot of cases, the host cannot really run without some of these running and it should fail to startup if an exception is thrown.
> 
> In fact, the only one that I think should have any try/catch around its startup is the `AssemblyAnalysisService`. I think we should backout the changes from all the others.

In the other cases (outside of AssemblyAnalysisService), is it worth logging the error in a catch statement and then throwing? Or are the logs not valuable here?

**Comment by brettsam** [comment]
> In the other cases (outside of AssemblyAnalysisService), is it worth logging the error in a catch statement and then throwing? Or are the logs not valuable here?

The logs will bubble up and prevent the Host from completing its startup. All of that will be logged automatically.

**Pull request closed without merging**